packages/liveblocks-node/src/client.ts

```diff
@@ -1891,6 +1971,188 @@ export class Liveblocks {
+    let opsBuffer: Op[] = [];
+    let outstandingFlush$: Promise<void> | undefined = undefined;
+    let lastFlush = performance.now();
+
+    const flushIfNeeded = (force: boolean) => {
+      if (opsBuffer.length === 0)
+        // Nothing to do
+        return;
+
+      if (outstandingFlush$) {
+        // There already is an outstanding flush, wait for it to complete
+        return;
+      }
+
+      const now = performance.now();
+      if (!(force || now - lastFlush > flushInterval)) {
+        // We're still within the debounce window, do nothing right now
+        return;
+      }
+
+      // All good, flush right now
+      lastFlush = now;
+      const ops = opsBuffer;
+      opsBuffer = [];
+
+      outstandingFlush$ = this.#sendMessage(
+        roomId,
+        [{ type: ClientMsgCode.UPDATE_STORAGE, ops }],
+        { signal }
+      ).finally(() => {
+        outstandingFlush$ = undefined;
+      });
+    };
+
+    // Download the storage contents
+    try {
+      const resp = await this.#startStorageMutation(roomId, { signal });
+      const { actor, nodes } = resp;
+
+      // Create a new pool
+      const pool = createManagedPool(roomId, {
+        getCurrentConnectionId: () => actor,
+        onDispatch: (
+          ops: Op[],
+          _reverse: Op[],
+          _storageUpdates: Map<string, StorageUpdate>
+        ) => {
+          if (ops.length === 0) return;
+
+          // Capture all the changes to the pool
+          for (const op of ops) {
+            opsBuffer.push(op);
+          }
+          flushIfNeeded(/* force */ false);
+        },
+      });
+
+      // Construct the Live tree
+      const root = LiveObject._fromItems(nodes, pool);
+
+      // Run the callback
+      await callback({ room, root: root as LiveObject<S> });
+    } catch (e) {
+      abort();
+      throw e;
+    } finally {
+      // Await any outstanding flushes, and then flush one last time
+      await outstandingFlush$; // eslint-disable-line @typescript-eslint/await-thenable
+      flushIfNeeded(/* force */ true);
+      await outstandingFlush$; // eslint-disable-line @typescript-eslint/await-thenable
+    }
+  }
+
+  async #sendMessage(
+    roomId: string,
+    messages: ClientMsg<JsonObject, Json>[],
+    options?: RequestOptions
+  ) {
+    const res = await this.#post(
+      url`/v2/rooms/${roomId}/send-message`,
+      { messages },
+      { signal: options?.signal }
+    );
+    if (!res.ok) {
+      throw await LiveblocksError.from(res);
+    }
+
+    // TODO: If res.ok, it will be a 200 response containing all returned Ops.
+    // These may include fix ops, which should get applied back to the managed
+    // pool.
+    // TODO Implement the handling of fix-ops:
+    // const data = (await res.json()) as {
+    //   messages: readonly (
+    //     | ServerMsg<JsonObject, BaseUserMeta, Json>
+    //     | readonly ServerMsg<JsonObject, BaseUserMeta, Json>[]
+    //   )[];
+    // };
+    // return data;
+  }
```

Comment:
I am curious how this would work. Would we update the `root` as we flush and if there are corrections to make to the storage?

Reply:
Yep, it would be the same as how the browser client works. The edge case this fixes for is small. I think 99% of users won't ever notice anything for this use case, but it's something that in theory (and thus eventually for some users) might be causing ordering issues when updating LiveLists from a mutation that are also simultaneously being edited by users.